omniscidb/QueryBuilder/QueryBuilder.cpp

```diff
@@ -0,0 +1,3072 @@
+      rhs.expr()->type()->isInterval()) {
+    return add(rhs.uminus());
+  }
+  try {
+    auto bin_oper = Analyzer::normalizeOperExpr(
+        OpType::kMinus, Qualifier::kOne, expr_, rhs.expr(), nullptr);
+    return {builder_, bin_oper, "", true};
+  } catch (std::runtime_error& e) {
+    throw InvalidQueryError() << "Cannot apply MINUS operation for operand types "
+                              << expr_->type()->toString() << " and "
+                              << rhs.expr()->type()->toString();
+  }
+}
+
+BuilderExpr BuilderExpr::sub(int val) const {
+  return sub(builder_->cst(val, builder_->ctx_.int32(false)));
+}
+
+BuilderExpr BuilderExpr::sub(int64_t val) const {
+  return sub(builder_->cst(val, builder_->ctx_.int64(false)));
+}
+
+BuilderExpr BuilderExpr::sub(float val) const {
+  return sub(builder_->cst(val, builder_->ctx_.fp32(false)));
+}
+
+BuilderExpr BuilderExpr::sub(double val) const {
+  return sub(builder_->cst(val, builder_->ctx_.fp64(false)));
+}
+
+BuilderExpr BuilderExpr::sub(const BuilderExpr& rhs, DateAddField field) const {
+  return add(rhs.uminus(), field);
+}
+
+BuilderExpr BuilderExpr::sub(int val, DateAddField field) const {
+  return sub(builder_->cst(val, builder_->ctx_.int32(false)), field);
+}
+
+BuilderExpr BuilderExpr::sub(int64_t val, DateAddField field) const {
+  return sub(builder_->cst(val, builder_->ctx_.int64(false)), field);
+}
+
+BuilderExpr BuilderExpr::sub(const BuilderExpr& rhs, const std::string& field) const {
+  return sub(rhs, parseDateAddField(field));
+}
+
+BuilderExpr BuilderExpr::sub(int val, const std::string& field) const {
+  return sub(builder_->cst(val, builder_->ctx_.int32(false)), parseDateAddField(field));
+}
+
+BuilderExpr BuilderExpr::sub(int64_t val, const std::string& field) const {
+  return sub(builder_->cst(val, builder_->ctx_.int64(false)), parseDateAddField(field));
+}
+
+BuilderExpr BuilderExpr::mul(const BuilderExpr& rhs) const {
+  try {
+    auto bin_oper = Analyzer::normalizeOperExpr(
+        OpType::kMul, Qualifier::kOne, expr_, rhs.expr(), nullptr);
+    return {builder_, bin_oper, "", true};
+  } catch (std::runtime_error& e) {
+    throw InvalidQueryError() << "Cannot apply MUL operation for operand types "
+                              << expr_->type()->toString() << " and "
+                              << rhs.expr()->type()->toString();
+  }
+}
+
+BuilderExpr BuilderExpr::mul(int val) const {
+  return mul(builder_->cst(val, builder_->ctx_.int32(false)));
+}
+
+BuilderExpr BuilderExpr::mul(int64_t val) const {
+  return mul(builder_->cst(val, builder_->ctx_.int64(false)));
+}
+
+BuilderExpr BuilderExpr::mul(float val) const {
+  return mul(builder_->cst(val, builder_->ctx_.fp32(false)));
+}
+
+BuilderExpr BuilderExpr::mul(double val) const {
+  return mul(builder_->cst(val, builder_->ctx_.fp64(false)));
+}
+
+BuilderExpr BuilderExpr::div(const BuilderExpr& rhs) const {
+  try {
+    auto bin_oper = Analyzer::normalizeOperExpr(
```

Comment:
Would it check for a constant div by 0?

Reply:
No. Normalization only makes sure we produce an expression we can handle in our codegen. It is mostly to add required casts to operands and choose the proper type for the result. I don't know if it makes sense to check for a division by zero because just having it is not an error (expression might be conditional and never execute).